### PR TITLE
[Enigmatrix] iP

### PR DESCRIPTION
**Pull request by Enigmatrix**
# Kude
> Subtracting from your list of priorities is as important as adding to it. - [Frank Sonnenberg](https://www.goodreads.com/quotes/tag/to-do-lists)

Kude feels like interacting with an unfeeling system that only cares about getting tasks done.
- Monotonic voice
- Straightforward
- Get ~~shit~~ tasks done quickly.

## How does it work?
1. Use `event` or `deadline` or `todo` to create tasks
2. Use `mark` or `unmark` to set tasks as done/undone
3. `list` or `find` tasks to do

What are you waiting for? Go and `TODO`! 😸 
```java
public class Kude {
    public static void main() {
         // TODO
    }
}
```
- [ ] TODO tick this item

### CODE REVIEW

File: src/main/java/kude/tui/Parser.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
+package kude.tui;
+
+import java.util.*;
```
Comment (bokung) [inline src/main/java/kude/tui/Parser.java]: Here perhaps the wildcard import can be avoided?

File: src/main/java/kude/tui/Processor.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+import kude.models.Deadline;
+import kude.models.Event;
+import kude.models.TaskList;
+import kude.models.Todo;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Optional;
+
+public class Processor {
```
Comment (bokung) [inline src/main/java/kude/tui/Processor.java]: Some documentation here would be very helpful!

File: src/main/java/kude/Duke.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
+package kude;
+
+import kude.tui.Processor;
+import kude.tui.Ui;
+
+public class Duke {
+
+    public static void main(String[] args) {
```
Comment (bokung) [inline src/main/java/kude/Duke.java:14]: Very well organised and clean main method!

File: src/main/java/kude/models/TaskList.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
+package kude.models;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+public class TaskList implements Serializable {
```
Comment (bokung) [inline src/main/java/kude/models/TaskList.java:12]: Good use of serialization for saving tasks!

File: src/main/java/kude/tui/Processor.java [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+        } catch (ClassNotFoundException cnfe) {
+            this.ui.writeErr("Save file corrupted or tampered with, reinitializing");
+        }
+        if (tasks == null) {
+            tasks = new TaskList();
+            saveTaskList();
+        }
+        this.tasks = tasks;
+    }
+
```
Comment (bokung) [inline src/main/java/kude/tui/Processor.java]: Beautiful organisation of commands!

File: src/main/java/kude/tui/Ui.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+    public void writeLine(String line) {
+        writeLine(line, 1);
+    }
+
+    public void writeLine(String line, int indent) {
+        out.printf("<%s%s%n", "\t".repeat(indent), line);
+    }
+
+    public void writeErr(String line) {
```
Comment (bokung) [inline src/main/java/kude/tui/Ui.java]: Might be a matter of personal preference here but can consider naming this method `writeError`! It's not much longer and maintains the consistency of using full-length words in this class!

**Review by bokung — COMMENTED** [review]
Very well done! Add some java docs and you are good to go!

File: src/main/java/kude/Duke.java (hunk 3 above)
Comment (Enigmatrix) [reply src/main/java/kude/Duke.java:14]: Thanks :)

File: src/main/java/kude/tui/Processor.java (hunk 5 above)
Comment (Enigmatrix) [reply src/main/java/kude/tui/Processor.java]: 👍

File: src/main/java/kude/tui/Ui.java (hunk 6 above)
Comment (Enigmatrix) [reply src/main/java/kude/tui/Ui.java]: 👍 🙇

File: src/main/java/kude/models/TaskList.java (hunk 4 above)
Comment (Enigmatrix) [reply src/main/java/kude/models/TaskList.java:12]: thanks :) 😸

**Pull request closed without merging**